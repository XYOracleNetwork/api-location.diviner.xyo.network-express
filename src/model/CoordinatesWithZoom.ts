import { Zoom } from './Zoom'

export type CoordinatesWithZoom = [number, number, Zoom]
