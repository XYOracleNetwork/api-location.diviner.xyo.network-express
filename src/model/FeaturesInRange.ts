import { XyoArchivistArchiveApi } from '@xyo-network/sdk-xyo-client-js'
import { Feature, GeoJsonProperties, Geometry } from 'geojson'

export type FeaturesInRange<G extends Geometry | null = Geometry, P = GeoJsonProperties> = (
  api: XyoArchivistArchiveApi,
  startTime: number,
  stopTime: number
) => Promise<Feature<G, P>[]>
