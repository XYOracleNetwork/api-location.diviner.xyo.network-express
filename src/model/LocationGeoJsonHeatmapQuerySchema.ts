import { LocationWitnessPayloadSchema } from '@xyo-network/sdk-xyo-client-js'

// TODO: Move to SDK
// TODO: Retype default heatmap query to use this schema as it's more explicit
// about the return format
export type LocationGeoJsonHeatmapQuerySchema = 'network.xyo.location.heatmap.geojson.query'
export const locationGeoJsonHeatmapQuerySchema: LocationGeoJsonHeatmapQuerySchema = 'network.xyo.location.heatmap.geojson.query'
export type LocationGeoJsonHeatmapAnswerSchema = 'network.xyo.location.heatmap.geojson.answer'
export const locationGeoJsonHeatmapAnswerSchema: LocationGeoJsonHeatmapAnswerSchema = 'network.xyo.location.heatmap.geojson.answer'
export type LocationGeoJsonHeatmapQuery = {
  startTime?: string
  stopTime?: string
  schema: LocationWitnessPayloadSchema
}
