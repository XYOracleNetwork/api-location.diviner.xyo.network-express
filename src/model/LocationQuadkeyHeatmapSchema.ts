import { LocationWitnessPayloadSchema } from '@xyo-network/sdk-xyo-client-js'

// TODO: Move to SDK
export type LocationQuadkeyHeatmapQuerySchema = 'network.xyo.location.heatmap.quadkey.query'
export const locationQuadkeyHeatmapQuerySchema: LocationQuadkeyHeatmapQuerySchema =
  'network.xyo.location.heatmap.quadkey.query'
export type LocationQuadkeyHeatmapAnswerSchema = 'network.xyo.location.heatmap.quadkey.answer'
export const locationQuadkeyHeatmapAnswerSchema: LocationQuadkeyHeatmapAnswerSchema =
  'network.xyo.location.heatmap.quadkey.answer'
export type LocationQuadkeyHeatmapQuery = {
  startTime?: string
  stopTime?: string
  schema: LocationWitnessPayloadSchema
}

export type LocationGeojsonHeatmapQuerySchema = 'network.xyo.location.heatmap.geojson.query'
export const locationGeojsonHeatmapQuerySchema: LocationGeojsonHeatmapQuerySchema =
  'network.xyo.location.heatmap.geojson.query'
export type LocationGeojsonHeatmapAnswerSchema = 'network.xyo.location.heatmap.geojson.answer'
export const locationGeojsonHeatmapAnswerSchema: LocationGeojsonHeatmapAnswerSchema =
  'network.xyo.location.heatmap.geojson.answer'
export type LocationGeojsonHeatmapQuery = {
  startTime?: string
  stopTime?: string
  schema: LocationWitnessPayloadSchema
}
