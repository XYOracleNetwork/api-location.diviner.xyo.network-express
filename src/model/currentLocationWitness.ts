export type CurrentLocationWitnessSchema = 'co.coinapp.currentlocationwitness'

export interface CurrentLocationWitness {
  _archive: string
  _hash: string
  _timestamp: number
  altitudeMeters: number
  directionDegrees: number
  latitude: number
  longitude: number
  quadkey: string
  schema: CurrentLocationWitnessSchema
  speedKph: number
}
