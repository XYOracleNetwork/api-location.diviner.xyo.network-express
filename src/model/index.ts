export * from './FeaturesInRange'
export * from './LocationAnswerSchema'
export * from './LocationQuadkeyHeatmapSchema'
export * from './SortDirection'
export * from './SupportedLocationSchemas'
export * from './TestData'
export * from './TileWithZoom'
export * from './WithHashPointProperties'
export * from './WithOptionalTimestamp'
export * from './WithTimestamp'
export * from './Zoom'
