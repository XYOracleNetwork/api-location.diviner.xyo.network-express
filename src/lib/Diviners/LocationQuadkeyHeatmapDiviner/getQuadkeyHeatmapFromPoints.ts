import { FeatureCollection, Point } from 'geojson'

import { QuadkeyWithDensity, WithHashProperties, Zoom } from '../../../model'
import {
  featureToQuadkey,
  getQuadkeyAtZoomLevel,
  getQuadkeysByParent,
  getQuadkeysByParentAtZoomLevel,
  getZoomLevel,
} from '../../Quadkey'

const minLocationsPerTile = 2
const minHeatmapZoom: Zoom = 9
const maxHeatmapZoom: Zoom = 12
const densityMultiplier = 1 / (minLocationsPerTile * maxHeatmapZoom)

// NOTE: Can we use numbers instead of strings for performance
const rollup = (quadkeys: string[], zoom: Zoom): string[] => {
  const locations = quadkeys.length
  if (zoom >= maxHeatmapZoom - 1 || locations < minLocationsPerTile) {
    // Base case, stop recursing
    // Convert keys to this zoom level (this is lossy and
    // that's what we want)
    return quadkeys.map((q) => getQuadkeyAtZoomLevel(q, zoom))
  } else {
    // Recursive case
    const nextZoom = (zoom + 1) as Zoom
    // Group by parent tile
    const quadkeysByParent = getQuadkeysByParentAtZoomLevel(quadkeys, zoom)
    // Rollup all the quadkeys at the next zoom level
    return Object.keys(quadkeysByParent)
      .map((q) => rollup(quadkeysByParent[q], nextZoom))
      .flatMap((q) => q)
  }
}

export const getQuadkeyHeatmapFromPoints = (
  points: FeatureCollection<Point, WithHashProperties>
): QuadkeyWithDensity[] => {
  const quadkeys = points.features
    // Calculate each point at max allowable zoom level
    .map<string>((p) => featureToQuadkey(p, maxHeatmapZoom))
  // Rollup quadkeys to an array starting at the minium zoom
  const heatmap = rollup(quadkeys, minHeatmapZoom)
  const quadkeysByParent = getQuadkeysByParent(heatmap)
  const quadkeysWithDensity = Object.keys(quadkeysByParent).map<QuadkeyWithDensity>((parent) => {
    const parentZoom = getZoomLevel(parent)
    return {
      // Normalize density to zoom level so that same number of points
      // at higher zoom is a higher density
      density: quadkeysByParent[parent].length * parentZoom * densityMultiplier,
      quadkey: parent,
    }
  })
  const maxCount = Math.max(...quadkeysWithDensity.map((q) => q.density))
  const quadkeysWithNormalizedDensity = quadkeysWithDensity.map((q) => {
    return {
      density: q.density / maxCount,
      quadkey: q.quadkey,
    }
  })
  return quadkeysWithNormalizedDensity
}
