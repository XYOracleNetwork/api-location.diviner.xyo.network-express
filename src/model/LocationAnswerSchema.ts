import { LocationHeatmapAnswerSchema, LocationQuadkeyHeatmapAnswerSchema, LocationTimeRangeAnswerSchema } from '@xyo-network/sdk-xyo-client-js'

import { LocationGeoJsonHeatmapAnswerSchema } from './LocationGeoJsonHeatmapQuerySchema'

// TODO: Move to SDK
export type LocationAnswerSchema = LocationTimeRangeAnswerSchema | LocationHeatmapAnswerSchema | LocationQuadkeyHeatmapAnswerSchema | LocationGeoJsonHeatmapAnswerSchema
