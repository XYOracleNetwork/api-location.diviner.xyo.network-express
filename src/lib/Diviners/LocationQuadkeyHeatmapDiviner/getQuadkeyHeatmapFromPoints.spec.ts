import { points } from '@turf/turf'
import { Position } from 'geojson'

import { TestData, WithHashProperties } from '../../../model'
import { getQuadkeyHeatmapFromPoints } from './getQuadkeyHeatmapFromPoints'

type TestDataInput = {
  coordinates: Position[]
}
type TestDataExpected = {
  quadkeys: number
}

const testData: TestData<TestDataInput, TestDataExpected>[] = [
  {
    expected: {
      quadkeys: 4,
    },
    input: {
      coordinates: [
        [-90, -45],
        [-90, 45],
        [90, -45],
        [90, 45],
      ],
    },
  },
  {
    expected: {
      quadkeys: 4,
    },
    input: {
      coordinates: [
        [-180, -90],
        [-180, -90],
        [-180, -90],
        [-180, -90],
      ],
    },
  },
]

describe('getQuadkeyHeatmapFromPoints', () => {
  it.each(testData)('calculates with known points', (data) => {
    const { coordinates } = data.input
    const locations = points<WithHashProperties>(coordinates, { hash: '' })
    expect(getQuadkeyHeatmapFromPoints(locations)).toMatchSnapshot()
  })
})
