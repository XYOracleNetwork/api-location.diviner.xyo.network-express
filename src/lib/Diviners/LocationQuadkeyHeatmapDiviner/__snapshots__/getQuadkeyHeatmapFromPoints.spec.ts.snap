// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getQuadkeyHeatmapFromPoints calculates with known points 1`] = `
Array [
  Object {
    "density": 1,
    "quadkey": "0",
  },
  Object {
    "density": 1,
    "quadkey": "1",
  },
  Object {
    "density": 1,
    "quadkey": "2",
  },
  Object {
    "density": 1,
    "quadkey": "3",
  },
]
`;

exports[`getQuadkeyHeatmapFromPoints calculates with known points 2`] = `
Array [
  Object {
    "density": 4,
    "quadkey": "0000000000000000000",
  },
]
`;
